### PR TITLE
Add support for Microsemi Mi-V softcore

### PR DESCRIPTION
**Pull request by kgugala**
This PR adds support for the Microsemi Mi-V soft core. The code has been tested on the M2GL025 IGLOO2 FPGA development board.

The PR adds three commits:

The first one disables the PLIC (RISC-V Platform Level Interrupt Controller) driver for the QEMU target. This is required because QEMU does not use PLIC. Without this change the second commit of this PR would not build for the QEMU target.

The second one moves the generic RISC-V code from the FE310 board to the common/ folder, so it can be reused by other platforms.

The third commit adds support for the Microsemi Mi-V soft CPU.

There are two things to discuss in this PR:

1. This PR does not add DTS support for the RISC-V boards. There is a PR open for the QEMU target which should be merged first (#7543)
2. We think that the `riscv-privilege` folder should be renamed to `riscv-privileged` as this is the correct name of this version of the RISC-V ISA (https://riscv.org/specifications/privileged-isa/). Nonetheless, the renaming is not a part of this PR, but could be discussed here.

### CODE REVIEW

File: boards/riscv32/m2gl025_miv/doc/m2g1025_miv.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+`Microsemi's SoftConsole website
+<https://www.microsemi.com/product-directory/design-tools/4879-softconsole>`_.
+
+With the necessary tools installed, you can connect to the board using OpenOCD.
+To establish an OpenOCD connection run:
+
+.. code-block:: bash
+
+   sudo LD_LIBRARY_PATH=<softconsole_path>/openocd/bin \
+   <softconsole install path>/openocd/bin/openocd  --file \
```
Comment (dbkinder) [inline boards/riscv32/m2gl025_miv/doc/m2g1025_miv.rst]: Should all three of these placeholders be ``<softconsole_path>``?  Not sure why the second one is different.

**Review by dbkinder — DISMISSED** [review]

File: drivers/serial/uart_miv.c [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+	((struct uart_miv_regs_t *)(DEV_CFG(dev))->uart_addr)
+#define DEV_DATA(dev)						\
+	((struct uart_miv_data * const)(dev)->driver_data)
+
+static unsigned char uart_miv_poll_out(struct device *dev,
+				       unsigned char c)
+{
+	volatile struct uart_miv_regs_t *uart = DEV_UART(dev);
+
+	while (!(uart->status & STATUS_TXRDY_MASK))
```
Comment (andrewboie) [inline drivers/serial/uart_miv.c]: wondering if there could be concurrency problems if two threads are attempting to poll out at the same time.

File: boards/riscv32/m2gl025_miv/doc/m2g1025_miv.rst (hunk 1 above)
Comment (kgugala) [reply boards/riscv32/m2gl025_miv/doc/m2g1025_miv.rst]: Thanks for catching that. It is now fixed.

File: drivers/serial/uart_miv.c (hunk 2 above)
Comment (kgugala) [reply drivers/serial/uart_miv.c]: Other serial drivers do not implement any critical section guards in the `poll_*` functions. Actually, the driver functions should not be called directly, but rather through a SYSCALL. The upper Zephyr layers should handle the concurrent accesses.

Comment (mgielda) [reply drivers/serial/uart_miv.c]: Hi @andrewboie is that an acceptable answer? Or do you have a concrete recommendation analogous to how it's done for some other platform? Reason I'm asking is that nobody's probably going to look at this PR when they see the "Changes requested" status ;)

Comment (mgielda) [reply drivers/serial/uart_miv.c]: Ah, I see it's @dbkinder's review that's the blocker here, so reiterating the request to him too!

File: boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+CONFIG_UART_MIV=y
+CONFIG_UART_MIV_PORT_0=y
+CONFIG_UART_MIV_PORT_0_BAUD_RATE=115200
+CONFIG_UART_MIV_PORT_0_NAME="uart0"
+CONFIG_UART_CONSOLE=y
+CONFIG_UART_CONSOLE_ON_DEV_NAME="uart0"
+CONFIG_PLIC=y
+CONFIG_RISCV_MACHINE_TIMER=y
+CONFIG_GPIO=n
+CONFIG_BOOT_BANNER=y
```
Comment (andrewboie) [inline boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig]: don't need BOOT_BANNER in the defconfig

**Review by tyrantlizard77 — APPROVED** [review]

**Review by pgielda — APPROVED** [review]

**Comment by mgielda** [comment]
Hi, I think we need @dbkinder to ACK us here and we're good to go?

**Review by nashif — APPROVED** [review]

File: boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig (hunk 3 above)
Comment (nashif) [reply boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig]: can you remove the boot banner flag from the defconfig?

**Review by dbkinder — APPROVED** [review]

File: boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig (hunk 3 above)
Comment (kgugala) [reply boards/riscv32/m2gl025_miv/m2gl025_miv_defconfig]: it is now removed.

**Pull request merged**